### ISSUE TITLE
Decoding MediaPlaylist, winsize = 0, Encode has no segments.

### ISSUE THREAD

**Issue by itsjamie** [grafov/m3u8#22]
When decoding a MediaPlaylist, and then reencoding it with a window size of zero. No segments are in the resulting file.

I traced it down to a the for loop inside Encode that loops the Segments, it currently checks `i < p.winsize`. Which works for a sliding playlist, but where VOD is supposed to set winsize to zero, it results in no segments being printed.

I was able to get rid of the issue by changing the for loop condition to be i <= p.winsize.. but I'm not sure if that effects live playlists in a poor way.

**Comment by bradleyfalzon** [comment]
@grafov, what's your thoughts on this?

It appears to create a NewMediaPlaylist for VoD, we need to set the window size to 0 and the capacity to some large number, then call the Decode or DecodeFrom on this newly created playlist.

Such as  (where `media.m3u8` is a copy of `sample-playlists/wowza-vod-chunklist.m3u8`):

```
    f, err := os.Open("media.m3u8")
    if err != nil {
        fmt.Println(err)
    }
    p, err := m3u8.NewMediaPlaylist(0, 1000)
    if err != nil {
        fmt.Println(err)
    }
    err = p.DecodeFrom(bufio.NewReader(f), false)
    if err != nil {
        fmt.Println(err)
    }

    fmt.Printf("Playlist object: %+v\n", p)
```

If someone follows the above steps, then the playlist will be empty but PR #38 should fix the problem.

But if someone uses the following method, where they might not know ahead of time whether the playlist is a Master or Media (or even if it's VoD or Live).

```
    f, err := os.Open("media.m3u8")
    if err != nil {
        panic(err)
    }
    p, listType, err := m3u8.DecodeFrom(bufio.NewReader(f), true)
    if err != nil {
        panic(err)
    }
    switch listType {
    case m3u8.MEDIA:
        mediapl := p.(*m3u8.MediaPlaylist)
        fmt.Printf("%+v\n", mediapl)
    case m3u8.MASTER:
        masterpl := p.(*m3u8.MasterPlaylist)
        fmt.Printf("%+v\n", masterpl)
    }
```

The window size appears to be set to 8 (see https://github.com/grafov/m3u8/blob/11af315b1d436b4b501d34b0c7b09323130a089f/reader.go#L155), and the full playlist is not shown (only 8 segments, not the full amount). I don't believe anyone can change the window size after the fact either.

So, PR #38 could be merged (I'd still like to do some additional testing though), but then we should modify the Decode methods in reader.go to set the `winsize` to 0 when playlist is Closed (when playlist is a VoD). For example:

```
--- a/reader.go
+++ b/reader.go
@@ -362,6 +362,7 @@ func decodeLineOfMediaPlaylist(p *MediaPlaylist, wv *WV, state *decodingState, l
        case line == "#EXT-X-ENDLIST":
                state.listType = MEDIA
                p.Closed = true
+               p.winsize = 0
        case strings.HasPrefix(line, "#EXT-X-VERSION:"):
                state.listType = MEDIA
                if _, err = fmt.Sscanf(line, "#EXT-X-VERSION:%d", &p.ver); strict && err != nil {
```

But because `decodeLineOfMediaPlaylist` is also used when a user may have explicitly called `NewMediaPlaylist(3, 1000)` the above proposal would overwrite the 3 to 0 - which may not be what the user wanted. So perhaps a better option would be to modify the Decode function to choose a different window size once the playlist has been decoded - if it's live set `winsize = 8`, if it's VoD, set to `winsize = 0`. So, I'm probably leaning towards the following:

```
--- a/reader.go
+++ b/reader.go
@@ -194,6 +194,9 @@ func decode(buf *bytes.Buffer, strict bool) (Playlist, ListType, error) {
        case MASTER:
                return master, MASTER, nil
        case MEDIA:
+               if media.Closed {
+                       media.winsize = 0
+               }
                return media, MEDIA, nil
        default:
                return nil, state.listType, errors.New("Can't detect playlist type")
```

Perhaps another PR could include options to overwrite the `winsize` as the decode function does set to 8, others (probably myself) might want to change this on the fly.

**Comment by bradleyfalzon** [comment]
I've proposed exposing the winsize as well as setting it correctly on VoD in #70.

**Comment by bradleyfalzon** [comment]
This should now be resolved in #70, let me know if it's not.
